### PR TITLE
Fix Ignite schema change operation tests failure

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description
Fix https://github.com/trinodb/trino/issues/16671

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

**Comment by ebyhr** [comment]
Could you change to draft PR and fix the title to clarify this PR isn't ready for review?

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/issues/16882 could be solved as well

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/4782527427/jobs/8502012841?pr=17102#step:3:43231 not related

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/4783122179/jobs/8504811292?pr=17102#step:3:45578 not related

**Comment by chenjian2664** [comment]
I will send the pr to solve the https://github.com/trinodb/trino/issues/16882, as we can see the above tests, there is no https://github.com/trinodb/trino/issues/16882 occurs after we upgrade the version to 2.9. cc @ebyhr

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestingIgniteContainer.java [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
     public TestingIgniteContainer()
     {
-        super(DockerImageName.parse("apacheignite/ignite:2.8.0"));
+        super(DockerImageName.parse("apacheignite/ignite:2.9.0"));
```
Comment (ebyhr) [inline plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestingIgniteContainer.java:31]: Please update `ignite.rst`.

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java [hunk 2]
```diff
@@ -109,6 +110,55 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
...
+        skipSchemaChangeOperationLockInterruptedError(() -> super.testAddAndDropColumnName(columnName, delimited));
+    }
+
+    @Override
+    protected String dataMappingTableName(String trinoTypeName)
+    {
+        return "test_data_mapping_smoke_" + trinoTypeName.replaceAll("[^a-zA-Z0-9]", "_") + randomNameSuffix();
+    }
+
+    private void skipSchemaChangeOperationLockInterruptedError(Runnable schemaChangeOperation)
```
Comment (ebyhr) [inline plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:143]: Why not use `@Flaky` annotation?

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestingIgniteContainer.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestingIgniteContainer.java:31]: Please update the version in `EnvMultinodeIgnite`.

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java [hunk 3]
```diff
@@ -109,6 +110,55 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
...
+    }
+
+    @Override
+    protected void testAddAndDropColumnName(String columnName, boolean delimited)
+    {
+        skipSchemaChangeOperationLockInterruptedError(() -> super.testAddAndDropColumnName(columnName, delimited));
+    }
+
+    @Override
+    protected String dataMappingTableName(String trinoTypeName)
```
Comment (ebyhr) [inline plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:138]: Why do you override this method?

File: plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java [hunk 4]
```diff
@@ -506,7 +506,7 @@ public Map<String, Object> getTableProperties(Connection connection, JdbcTableHa
         String schemaName = requireNonNull(schemaTableName.getSchemaName(), "Ignite schema name can not be null").toUpperCase(ENGLISH);
         String tableName = requireNonNull(schemaTableName.getTableName(), "Ignite table name can not be null").toUpperCase(ENGLISH);
         // Get primary keys from 'sys.indexes' because DatabaseMetaData.getPrimaryKeys doesn't work well while table being concurrent modified
```
Comment (ebyhr) [inline plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:508]: Is this comment still valid even after upgrading the server version?

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java (hunk 2 above)
Comment (chenjian2664) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:143]: I saw the annotation is marked with `Deprecated`, shall we use it?

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java (hunk 3 above)
Comment (chenjian2664) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:138]: Ignite can not parse the table name that has two consecutive wildcard. I filed a [pr](https://github.com/apache/ignite/pull/10579) to fix  but still ongoing.

Comment (ebyhr) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:138]: Let's fix the base method instead.

File: plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java (hunk 4 above)
Comment (chenjian2664) [reply plugin/trino-ignite/src/main/java/io/trino/plugin/ignite/IgniteClient.java:508]: Yes, tested on my local, the `DatabaseMetaData.getPrimaryKeys` would return empty primary keys sometimes.

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java [hunk 5]
```diff
@@ -109,6 +110,49 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
...
+
+    private void skipSchemaChangeOperationLockInterruptedError(Runnable schemaChangeOperation)
+    {
+        try {
+            schemaChangeOperation.run();
+        }
+        catch (Exception e) {
+            RuntimeException trinoException = getTrinoExceptionCause(e);
+            try {
+                verifyConcurrentAddColumnFailurePermissible(trinoException);
```
Comment (findepi) [inline plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:145]: this is called in `testDropColumn` as well

Comment (chenjian2664) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:145]: Yes, we override the `testDropColumn` test here, any concern?

File: plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java [hunk 6]
```diff
@@ -29,6 +29,7 @@
```
Comment (kokosing) [inline plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:29]: What is the problem that you are fixing? Is this about concurrent access of ignite? Maybe we should run this test in serial way. WDYT?

Another idea is that we could potentially use `@Flaky` if the test can recover from the failure.

Comment (chenjian2664) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:29]: > What is the problem that you are fixing? Is this about concurrent access of ignite? Maybe we should run this test in serial way. WDYT?

Yes, the concurrent request for the "add" or "drop" column. 

> Maybe we should run this test in serial way.

I feel some tasks is tend to test the concurrent scenario about the the metadata request, for those tests maybe we shouldn't do in serial way. For the tests that not tend to test concurrent scenario maybe we can. 
And could we add a behavior like `SUPPORT_CONCURRENT_ON_COLUMN_MODIFICATION`

@Flaky is marked as Deprecated

Comment (kokosing) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:29]: > `@flaky` is marked as Deprecated

The point of this to discourage the usage. So people try to fix it. However here, if the issue is on the ignite side I think usage of that annotation is justified.

Comment (kokosing) [reply plugin/trino-ignite/src/test/java/io/trino/plugin/ignite/TestIgniteConnectorTest.java:29]: > SUPPORT_CONCURRENT_ON_COLUMN_MODIFICATION

That sounds like an overkill to me.

**Review by findepi — APPROVED** [review]

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Thank you!

**Pull request merged**